src/vcf/validator.cpp

```diff
@@ -150,38 +164,79 @@ namespace ebi
             return false;
         }
         std::unique_ptr<Parser> validator = build_parser(sourceName, validationLevel, version);
-        return validate(line, input, *validator, outputs);
+        return validate(line, uncompressed_input, *validator, outputs);
+    }
+
+    void create_uncompressed_stream(std::istream & input,
+                                    const std::string & sourceName,
+                                    boost::iostreams::filtering_istream & uncompressed_input)
+    {
+        std::vector<char> line;
+        get_magic_num(input, line);
+
+        std::string file_ext = get_compression(sourceName, line);
+        check_readability_of_file(file_ext);
+
+        if(file_ext == BZ2) {
+            uncompressed_input.push(boost::iostreams::bzip2_decompressor());
+        } else if(file_ext == TAR_GZ) {
+            uncompressed_input.push(boost::iostreams::gzip_decompressor());
+        }
+
+        uncompressed_input.push(input);
     }
 
-    bool is_compressed_file(const std::string &source,
-                            const std::vector<char> &line)
+    void get_magic_num(std::istream & stream, std::vector<char> & container)
     {
-        if (source != ebi::vcf::STDIN && is_compressed_extension(source)) {
-            return true;
+        char c;
+        int i=0;
+        container.clear();
+
+        while (stream && stream.get(c)) {
+            container.push_back(c);
+            i++;
+            if (c == '\n' || i > 4) break;
+        }
+
+        for (int j = 0; j < (int)container.size(); ++j) {
+            stream.unget();
+            if (stream.fail() || !stream) {
+                throw std::runtime_error("Stream failed");
```

Comment:
Add more context to this message. Imagine when will this message show up.

hypothetical situation in 12 months in the future:
> Ok I refactored the main.cpp and the validator.cpp a bit, but all tests should work. Oh they failed, and any validation is only showing:
```
$ cat my.vcf.gz |  ./vcf_validator 
Stream failed
$ 
```